File: civitasJava/src/civitas/JugadorEspeculador.java
/**
 * @file JugadorEspeculador.java
 * @author Yesenia González Dávila
 * @author Esther García Gallego
 * @note Grupo B.3
 * @class JugadorEspeculador
 * @brief Representa a un tipo de jugador especial
*/


package civitas;
public class JugadorEspeculador extends Jugador {

    private final int FactorEspeculador = 2;
    private float fianza;

    JugadorEspeculador(Jugador otro, float _fianza) {   
        super(otro);
        fianza =_fianza;
        
        for(int i = 0; i < getPropiedades().size(); i++)
            getPropiedades().get(i).ActualizaPropietarioPorConversion(this);
    }

    @Override
    protected boolean debeSerEncarcelado() {
        boolean carcel = false;

        if (!isEncarcelado()) {
            
            if (!tieneSalvoconducto())      // 1. Si no tiene salvoconducto
                
                if (puedoGastar(fianza)){   // 1.2. Puede pagar fianza
                    paga(fianza);
                    Diario.getInstance().ocurreEvento("El jugador especulador " + getNombre()
                            + " se ha librado de la cárcel pagando la fianza");
                    
                } else carcel = true;       // 1.3. Si no puede pagar fianza, se encarcela
                    
                
            else {                          // 2. Si tiene salvoconducto, lo pierde y no es encarcelado
                perderSalvoconducto();
                Diario.getInstance().ocurreEvento("El jugador especulador " + getNombre()
                        + " se ha librado de la cárcel por tener un salvoconducto");
            }
        }

        return carcel;
    }

    @Override
    boolean pagaImpuesto(float cantidad) {
        if (isEncarcelado()) 
            return false;
        else 
            return paga(cantidad / FactorEspeculador);
    }
   
    @Override 
    public String toString(){
        String s = super.toString() + " Especulador";
        return s;
    }
    
    @Override
    protected int getCasasMax(){
        return super.getCasasMax()*FactorEspeculador;
    }
    
    @Override
    protected int getHotelesMax () {
      return super.getHotelesMax()*FactorEspeculador;
    }
}

File: civitasRuby/lib/civitas/jugador_especulador.rb
#encoding:utf-8

=begin
Authors: Esther García Gallego
         Yesenia González Dávila
         Grupo B3
=end

module Civitas

class JugadorEspeculador < Jugador
  
  @@FactorEspeculador = 2 # Atributo de clase
    
  
  def initialize (jugador, fianza)
    super(jugador.nombre)
    @fianza = fianza
    nuevoEspeculador(jugador)
  end
  
  
  def nuevoEspeculador (jugador) 
    @saldo = jugador.saldo
    @encarcelado = jugador.isEncarcelado
    @puedeComprar = jugador.puedeComprar
    @numCasillaActual = jugador.numCasillaActual
    
    @propiedades = jugador.propiedades
    
    @propiedades.each do |propiedad| 
      propiedad.actualizaPropietarioPorConversion(self)
    end
  end
  
  
  def pagaImpuesto(cantidad)
    if (isEncarcelado)
      false
    else
      paga(cantidad/@@FactorEspeculador)
    end
  end
  
  
  def to_s
     str = " >> Jugador #{@nombre}. #{@saldo} €. Propiedades: #{@propiedades.length}. Edificaciones #{cantidadCasasHoteles}. "
    str = str + "\n    Casilla actual: #{@numCasillaActual}."
    str = str + "\n    Especulador."
    
    if @puedeComprar
      str = str + " Puede comprar."
    end
    if @encarcelado
      str = str + " Encarcelado."
    end
    if !@salvoconducto.nil?
      str = str + " Tiene salvoconducto."
    end
    
    str = str + "\n"
    
    str
  end
  
  private # ------------------------------------------------------------------ #
  
  def casasMax
    @CasasMax*@@FactorEspeculador
  end
  
  def hotelesMax
    @HotelesMax*@@FactorEspeculador
  end
  
  protected # ---------------------------------------------------------------- #
  
   def debeSerEncarcelado ()
    carcel = false
    
      if !isEncarcelado()
        
        if !tieneSalvoconducto       # 1. Si no tiene salvoconducto
          
          if puedoGastar(@fianza)    #     1.2 Puede pagar fianza y no es encarcelado
            paga(@fianza)
            Diario.instance.ocurre_evento ("El jugador especulador #{@nombre} se ha librado de la carcel pagando la fianza (#{@fianza}€)")
          else                       #     1.3. Se encarcela
            carcel = true
          end
          
         else                        # 2. Pierde el salvoconducto y no es encarcelado
          perderSalvoconducto()
          Diario.instance.ocurre_evento ("El jugador #{@nombre} se ha librado de la cárcel por tener un salvoconducto")
        end
      end
      
    carcel
  end
  
end

end
